ui/src/app/utils/cluster-initializer.spec.tsx

```diff
@@ -0,0 +1,305 @@
+
+    expect(mockDeleteCluster).toHaveBeenCalled();
+    expect(mockCreateCluster).toHaveBeenCalled();
+    expect(cluster.status).toEqual(ClusterStatus.Running);
+  });
+
+  it('Should recover from intermittent 500s', async() => {
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Creating},
+      {...baseCluster, status: ClusterStatus.Creating},
+    ]);
+    mockGetCluster.mockImplementationOnce(async(workspaceNamespace) => {
+      throw new Response(null, {status: 503});
+    });
+    mockGetCluster.mockImplementationOnce(async(workspaceNamespace) => {
+      throw new Response(null, {status: 503});
+    });
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Running},
+    ]);
+
+    const initializer = createInitializer();
+    const cluster = await runInitializerAndTimers(initializer);
+
+    expect(cluster.status).toEqual(ClusterStatus.Running);
+  });
+
+  it('Should give up after too many server errors', async() => {
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Creating},
+    ]);
+    for (let i = 0; i < 10; i++) {
+      mockGetCluster.mockImplementationOnce(async(workspaceNamespace) => {
+        throw new Response(null, {status: 503});
+      });
+    }
+
+    const initializer = createInitializer();
+
+    expect.assertions(1);
+    return runInitializerAndTimers(initializer).catch(error => {
+      expect(error.message).toMatch(/max server error count/i);
+    });
+  });
+
+    it('Should timeout after max delay', async() => {
+    mockGetCluster.mockImplementation(async(workspaceNamespace) => {
+      return {...baseCluster, status: ClusterStatus.Starting};
+    });
+
+    // There's some nuance / awkwardness to this test: the ClusterInitializer uses Date.now() to get
+    // the current timestamp, but Jest doesn't support fake clock functionality (see
+    // https://github.com/facebook/jest/issues/2684). So we just set a very quick timeout here to
+    // ensure the threshold is reached after a couple polling loops.
+    const initializer = createInitializer({overallTimeout: 30});
+
+    // Tell Jest that we plan to have 1 assertion. This ensures that the test won't
+    // pass if the promise fails.
+    expect.assertions(1);
+    return runInitializerAndTimers(initializer).catch(error => {
+      expect(error.message).toMatch(/max time allowed/i);
+    });
+  });
+
+  it('Should reject promise after abort signal', async() => {
+    mockGetCluster.mockImplementation(async(workspaceNamespace) => {
+      return {...baseCluster, status: ClusterStatus.Starting};
+    });
+    const aborter = new AbortController();
+    const initializer = createInitializer({
+      abortSignal: aborter.signal
+    });
+
+    const initializePromise = runInitializerAndTimers(initializer);
+
+    // Wait a reasonably-short amount of time, at least one polling delay period, before sending
+    // an abort signal.
+    await new Promise(resolve => setTimeout(resolve, 20));
```

Comment:
Maybe make a helper out of these two lines, like a `doAfter(fn, timeoutMillis)`?

Reply:
It's not obfuscating if you kill 4 keywords and 2 scopes in a one-liner: `{await, new, promise, setTimeout, fat arrow}`. It's only easy to read if you like reading JS. 😉